http://bugs.gentoo.org/245370

ripped from ncurses-5.7-20081213.patch

20081213
	+ add check for failure to open hashed-database needed for db4.6
	  (GenToo #245370).

--- ncurses-5.7-20081206+/ncurses/tinfo/hashed_db.c	2006-08-19 19:48:38.000000000 +0000
+++ ncurses-5.7-20081213/ncurses/tinfo/hashed_db.c	2008-12-13 20:59:02.000000000 +0000
@@ -49,27 +49,30 @@
 _nc_db_open(const char *path, bool modify)
 {
     DB *result = 0;
+    int code;
 
 #if HASHED_DB_API >= 4
     db_create(&result, NULL, 0);
-    result->open(result,
-		 NULL,
-		 path,
-		 NULL,
-		 DB_HASH,
-		 modify ? DB_CREATE : DB_RDONLY,
-		 0644);
+    if ((code = result->open(result,
+			     NULL,
+			     path,
+			     NULL,
+			     DB_HASH,
+			     modify ? DB_CREATE : DB_RDONLY,
+			     0644)) != 0) {
+	result = 0;
+    }
 #elif HASHED_DB_API >= 3
     db_create(&result, NULL, 0);
-    result->open(result,
-		 path,
-		 NULL,
-		 DB_HASH,
-		 modify ? DB_CREATE : DB_RDONLY,
-		 0644);
+    if ((code = result->open(result,
+			     path,
+			     NULL,
+			     DB_HASH,
+			     modify ? DB_CREATE : DB_RDONLY,
+			     0644)) != 0) {
+	result = 0;
+    }
 #elif HASHED_DB_API >= 2
-    int code;
-
     if ((code = db_open(path,
 			DB_HASH,
 			modify ? DB_CREATE : DB_RDONLY,
@@ -77,21 +80,22 @@
 			(DB_ENV *) 0,
 			(DB_INFO *) 0,
 			&result)) != 0) {
-	T(("cannot open %s: %s", path, strerror(code)));
 	result = 0;
-    } else {
-	T(("opened %s", path));
     }
 #else
-    result = dbopen(path,
-		    modify ? (O_CREAT | O_RDWR) : O_RDONLY,
-		    0644,
-		    DB_HASH,
-		    NULL);
+    if ((result = dbopen(path,
+			 modify ? (O_CREAT | O_RDWR) : O_RDONLY,
+			 0644,
+			 DB_HASH,
+			 NULL)) == 0) {
+	code = errno;
+    }
+#endif
     if (result != 0) {
 	T(("opened %s", path));
+    } else {
+	T(("cannot open %s: %s", path, strerror(code)));
     }
-#endif
     return result;
 }
 
